subsys/zigbee/osif/zb_nrf_serial.c

```diff
@@ -0,0 +1,467 @@
+			    &(uart_rx_buf[ZIGBEE_UART_RX_CHUNK_SIZE * 2]),
+			    uart_rx_buf_len - ZIGBEE_UART_RX_CHUNK_SIZE * 2);
+
+			/*
+			 * Async: Remember the free RX chunk and restore it once
+			 *        user's RX buffer will be filled with data.
+			 */
+			blocked_buf_chunk = uart_rx_buf_chunk_released;
+		} else if (uart_rx_buf_chunk_released) {
+			/* Sync: Pass the most recently released buffer. */
+			uart_rx_buf_rsp(uart_dev,
+					uart_rx_buf_chunk_released,
+					ZIGBEE_UART_RX_CHUNK_SIZE);
+		}
+
+		/* The RX chunk was either remembered or used for reception. */
+		uart_rx_buf_chunk_released = NULL;
+		break;
+
+	case UART_RX_BUF_RELEASED:
+		if (uart_rx_buf) {
+			store_rx_buf(evt->data.rx_buf.buf);
+			if (uart_rx_buf_offset == uart_rx_buf_len) {
+				uart_rx_buf_len = 0;
+				zigbee_schedule_callback(uart_rx_notify, 0);
+			}
+		}
+
+		/*
+		 * Pass the consumed buffer through uart_rx_buf_chunk_released
+		 * variable to keep RX chunks in a loop.
+		 */
+		if (!uart_rx_buf_chunk_released) {
+			if ((evt->data.rx_buf.buf == uart_rx_buf_chunk[0]) ||
+			    (evt->data.rx_buf.buf == uart_rx_buf_chunk[1])) {
+				uart_rx_buf_chunk_released =
+					evt->data.rx_buf.buf;
+			} else if (blocked_buf_chunk) {
+				uart_rx_buf_chunk_released = blocked_buf_chunk;
+				blocked_buf_chunk = NULL;
+			}
+		}
+		break;
+
+	case UART_RX_DISABLED:
+		if (restart_rx) {
+			/*
+			 * Async: Reception was stopped due to RX timeout,
+			 *        restart it automatically.
+			 */
+			restart_rx = false;
+			uart_rx_buf_chunk_released = uart_rx_buf_chunk[1];
+			(void)uart_rx_enable(uart_dev,
+					     uart_rx_buf_chunk[0],
+					     ZIGBEE_UART_RX_CHUNK_SIZE,
+					     CONFIG_ZIGBEE_UART_RX_TIMEOUT);
+		} else {
+			/*
+			 * Reception finished and the next RX operation may
+			 * be started.
+			 */
+			k_sem_give(&rx_done_sem);
+		}
+		break;
+
+	case UART_RX_STOPPED:
+		/* An error occurred during reception. Abort RX. */
+		if (rx_data_cb && uart_rx_buf) {
+			if (blocked_buf_chunk) {
+				uart_rx_buf_chunk_released = blocked_buf_chunk;
+				blocked_buf_chunk = NULL;
+			}
+			uart_rx_buf_len = 0;
+			zigbee_schedule_callback(uart_rx_notify, 0);
+		}
+		break;
+	default:
+		break;
+	}
+}
+
+
+void zb_osif_serial_init(void)
+{
+	if (uart_dev != NULL) {
+		return;
+	}
+
+	char_handler = NULL;
+	rx_data_cb = NULL;
+	tx_data_cb = NULL;
+	uart_tx_buf = uart_tx_buf_mem;
+	uart_tx_buf_len = sizeof(uart_tx_buf_mem);
+	uart_rx_buf = NULL;
+	uart_rx_buf_len = 0;
+	uart_rx_buf_offset = 0;
+
+	uart_dev = device_get_binding(CONFIG_ZIGBEE_UART_DEVICE_NAME);
+	if (uart_dev == NULL) {
+		return;
+	}
+
+	if (uart_callback_set(uart_dev, uart_evt_handler, NULL)) {
+		uart_dev = NULL;
+		return;
+	}
+
+	(void)k_sem_take(&rx_done_sem, K_FOREVER);
+	uart_rx_buf_chunk_released = uart_rx_buf_chunk[1];
+	if (uart_rx_enable(uart_dev,
+			   uart_rx_buf_chunk[0],
+			   ZIGBEE_UART_RX_CHUNK_SIZE,
+			   CONFIG_ZIGBEE_UART_RX_TIMEOUT)) {
+		uart_dev = NULL;
+		return;
```

Comment:
And what about `rx_done_sem`? Should it remain taken? Another call to `zb_osif_serial_init` would hang then.

Reply:
Fixed.